app/src/androidTest/java/com/google/jetpackcamera/utils/UiTestUtil.kt

```diff
@@ -77,3 +90,129 @@ fun ComposeTestRule.getCurrentLensFacing(): LensFacing {
+) :
+    TestRule {
+    private lateinit var wrappedRule: GrantPermissionRule
+
+    override fun apply(base: Statement, description: Description): Statement {
+        for (targetName in targetTestNames) {
+            if (description.methodName == targetName) {
+                wrappedRule = GrantPermissionRule.grant(*permissions)
+                return wrappedRule.apply(base, description)
+            }
+        }
+        // If no match, return the base statement without granting permissions
+        return base
+    }
+}
+
+// functions for interacting with system permission dialog
+@SdkSuppress(minSdkVersion = 30)
+fun askEveryTimeDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 30) {
+        Log.d(TAG, "Searching for Allow Once Button...")
+
+        val askPermission = findObjectById(
+            uiDevice = uiDevice,
+            resId = "com.android.permissioncontroller:id/permission_allow_one_time_button"
+        )
+
+        Log.d(TAG, "Clicking Allow Once Button")
+
+        askPermission!!.click()
+    }
+}
+
+/**
+ *  Clicks ALLOW option on an open permission dialog
+ */
+fun grantPermissionDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 23) {
+        Log.d(TAG, "Searching for Allow Button...")
+
+        val allowPermission = findObjectById(
+            uiDevice = uiDevice,
+            resId = when {
+                Build.VERSION.SDK_INT <= 29 ->
+                    "com.android.packageinstaller:id/permission_allow_button"
+                else -> "com.android.permissioncontroller:id/permission_allow_button"
+            }
+        )
+        Log.d(TAG, "Clicking Allow Button")
+
+        allowPermission!!.click()
+    }
+}
+
+/**
+ * Clicks the DENY option on an open permission dialog
+ */
+fun denyPermissionDialog(uiDevice: UiDevice) {
+    if (Build.VERSION.SDK_INT >= 23) {
+        Log.d(TAG, "Searching for Deny Button...")
+        val denyPermission = findObjectById(
+            uiDevice = uiDevice,
+
+            resId = when {
+                Build.VERSION.SDK_INT <= 29 ->
+                    "com.android.packageinstaller:id/permission_deny_button"
+                else -> "com.android.permissioncontroller:id/permission_deny_button"
+            }
+        )
+        Log.d(TAG, "Clicking Deny Button")
+
+        denyPermission!!.click()
+    }
+}
+
+/**
+ * Finds a system button by its resource ID.
+ * fails if not found
+ */
+fun findObjectById(
+    uiDevice: UiDevice,
+    resId: String,
+    timeout: Long = 10000,
+    shouldFailIfNotFound: Boolean = true
+): UiObject2? {
+    val selector = By.res(resId)
+    return if (!uiDevice.wait(Until.hasObject(selector), timeout)) {
+        if (shouldFailIfNotFound) {
+            fail("Could not find object with RESOURCE ID: $resId")
+        }
+        null
+    } else {
+        uiDevice.findObject(selector)
+    }
+}
+
+/**
+ * Finds a system button by its string value.
+ * fails if not found
+ */
+fun findObjectByText(
```

Comment:
nit: private

Reply:
unused, so removed it. finding by text can cause issues if the device is in another language, so best to stick with the id's